### ISSUE TITLE
MSM performance benefits

### ISSUE THREAD

**Issue by waamm** [supranational/sppark#54]
Hi! After quickly replacing the `VariableBaseMSM::multi_scalar_mul` [here](https://github.com/alex-ozdemir/collaborative-zksnark/blob/main/groth16/src/prover.rs) with `msm_cuda::multi_scalar_mult_arkworks`, I am not noticing any performance difference. Does that make sense?

**Comment by dot-asm** [comment]
> Seems to work now for some reason. Is there a specific reason this library uses v0.3 of arkworks?

This is not true. This **library** has literally no dependencies (build-dependencies don't count). Arkworks is used in the test suite purely for verification purposes.

**Comment by dot-asm** [comment]
> After quickly replacing the `VariableBaseMSM::multi_scalar_mul` invocations [here](https://github.com/alex-ozdemir/collaborative-zksnark/blob/main/groth16/src/prover.rs) with `msm_cuda::multi_scalar_mult_arkworks`, I am not noticing any performance difference.

The trouble is that any "quick" thing is more likely to fail to deliver the expected improvement. At least based on what we've seen. Orchestrating the data flow and more tight integration with application is the key. See https://github.com/supranational/supra_seal/tree/main/c2 for an example. Well, "see" is a bit of a misnomer, because you're unlikely to figure it out just like that. The keyword is rather that it takes over a bigger section of the prover in order to to the "orchestrating and tight integration" thing.

**Comment by dot-asm** [comment]
Though there is another caveat. Current MSM implementation is oversensitive to repeating bit patterns in scalars. I mean if scalars have some bit structure, and some provers were observed to produce more repetitive patterns than others, current MSM implementation would tend to underperform. The reason for why the issue is not resolved is because there was no strong motivating factor so far...

**Comment by dot-asm** [comment]
> See https://github.com/supranational/supra_seal/tree/main/c2 for an example.

Just in case, this is Groth16, a demanding one, operating on vectors with ~2\*\*27 elements. And the payoff was absolutely worth the effort.
